File: CCO '15 P2 - Artskjid.py
# NP problem go brr
# 48/50 using dp, check c++ code for 50/50 solution
from collections import defaultdict

n_nodes, n_roads = map(int, input().split())
graph = [[] for _ in range(n_nodes)]

for _ in range(n_roads):
    a, b, dist = map(int, input().split())
    graph[a].append((b, dist))

res = 0
max_dists = defaultdict(lambda: defaultdict(int))

stack = [(0, 1, 0)]
while stack:
    node, path, dist = stack.pop()
    if node == n_nodes - 1:
        res = max(res, dist)
        continue

    for adj, adj_dist in graph[node]:
        new_dist = dist + adj_dist
        if path & (1 << adj):
            continue

        new_path = path | (1 << adj)

        if new_dist > max_dists[adj][new_path]:
            max_dists[adj][new_path] = new_dist
            stack.append((adj, new_path, new_dist))

print(res)


# # 43/50 test cases, TLE
# n_nodes, n_roads = map(int, input().split())
# graph = [[] for _ in range(n_nodes)]
#
# for _ in range(n_roads):
#     a, b, dist = map(int, input().split())
#     graph[a].append((b, dist))
#
# max_dist = 0
# state = [False for _ in range(n_nodes)]
# state[0] = True
#
# memo = {}  # keep track of already visited paths
# q = [(0, tuple(state), 0)]
# while q:
#     node, path, dist = q.pop()
#     if node == n_nodes - 1:
#         max_dist = max(max_dist, dist)
#         continue
#
#     for adj, adj_dist in graph[node]:
#         new_dist = dist + adj_dist
#         if path[adj]:
#             continue
#
#         new_path = list(path)
#         new_path[adj] = True
#         new_path = tuple(new_path)
#
#         if (adj, new_path) not in memo or new_dist > memo[(adj, new_path)]:
#             memo[(adj, new_path)] = new_dist
#             q.append((adj, new_path, new_dist))
#
# print(max_dist)

File: CCO '15 P2 - Artskjid.cpp
// my python code was too slow
// passes 50/50 test cases

#include <iostream>
#include <vector>
#include <bitset>
#include <array>

using namespace std;

const int MAXN = 18;
int n_nodes, n_roads;
vector<vector<pair<int, int>>> graph;
int max_dist = 0;
bitset<MAXN> state;
array<array<int, 1 << MAXN>, MAXN> max_dists;

void dfs(int node, bitset<MAXN> &path, int dist) {
    if (node == n_nodes - 1) {
        max_dist = max(max_dist, dist);
        return;
    }

    for (auto &[adj, adj_dist] : graph[node]) {
        int new_dist = dist + adj_dist;
        if (path[adj]) continue;

        bitset<MAXN> new_path = path;
        new_path[adj] = true;

        if (!max_dists[adj][new_path.to_ullong()] || new_dist > max_dists[adj][new_path.to_ullong()]) {
            max_dists[adj][new_path.to_ullong()] = new_dist;
            dfs(adj, new_path, new_dist);
        }
    }
}

int main() {
    cin >> n_nodes >> n_roads;
    graph.resize(n_nodes);

    for (int i = 0; i < n_roads; i++) {
        int a, b, dist;
        cin >> a >> b >> dist;
        graph[a].push_back({b, dist});
    }

    state[0] = true;

    dfs(0, state, 0);
    cout << max_dist << endl;

    return 0;
}
